### PR TITLE
Speed up polynomial computation while improve Chunk speed a bit

### PR DESCRIPTION
**Pull request by wjiec**
Hi, I've recently referred to this pkg and related papers to learn about CDC, and in the process I've made some optimizations for polynomial computation. I also did some minor optimizations for Chunk in the process, which probably had a ~3% speedup:
```plain
goos: linux
goarch: amd64
pkg: github.com/restic/chunker
cpu: Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz
                    │    old.txt    │                new.txt                 │
                    │    sec/op     │    sec/op      vs base                 │
ChunkerWithSHA256-8    188.1m ±  0%    184.5m ±  0%    -1.89% (p=0.000 n=10)
Chunker-8              77.78m ±  0%    74.29m ±  0%    -4.49% (p=0.000 n=10)
NewChunker-8          102.31µ ± 23%    99.77µ ± 21%         ~ (p=0.353 n=10)
PolMul-8               195.9n ±  0%    127.5n ±  0%   -34.87% (p=0.000 n=10)
PolDiv-8               86.88n ±  0%    86.79n ±  0%    -0.10% (p=0.049 n=10)
PolDeg-8              0.3744n ±  0%   0.7498n ±  0%  +100.29% (p=0.000 n=10)
PolMod-8               86.89n ±  0%    86.80n ±  0%         ~ (p=0.578 n=10)
PolDivMod-8            86.29n ±  0%    86.30n ±  0%         ~ (p=0.540 n=10)
RandomPolynomial-8    2471.3µ ±  3%    377.7µ ±  1%   -84.72% (p=0.000 n=10)
PolIrreducible-8      1372.1µ ±  0%    130.1µ ±  0%   -90.52% (p=0.000 n=10)
PolGCD-8               789.5n ±  0%    738.2n ±  0%    -6.50% (p=0.000 n=10)
PolMulMod-8           9995.0n ±  0%    555.0n ±  0%   -94.45% (p=0.000 n=10)
geomean                10.56µ          5.881µ         -44.28%

                    │   old.txt    │               new.txt               │
                    │     B/s      │     B/s       vs base               │
ChunkerWithSHA256-8   170.1Mi ± 0%   173.4Mi ± 0%  +1.93% (p=0.000 n=10)
Chunker-8             411.4Mi ± 0%   430.8Mi ± 0%  +4.70% (p=0.000 n=10)
geomean               264.6Mi        273.3Mi       +3.30%
```

These are my test results in my homlab, it may not be accurate, if you have time please double check it. Meanwhile, I tested compiling restic with this optimized version and the program works fine(🤔️).

### CODE REVIEW

File: chunker.go [hunk 1]
```diff
@@ -131,11 +132,9 @@ func (c *BaseChunker) fillTables() {
 	//
 	// Afterwards a new byte can be shifted in.
 	for b := 0; b < 256; b++ {
-		var h Pol
-
-		h = appendByte(h, byte(b), c.pol)
-		for i := 0; i < windowSize-1; i++ {
-			h = appendByte(h, 0, c.pol)
+		h := Pol(b)
```
Comment (MichaelEischer) [inline chunker.go:135]: I'm not sure whether it's worth the trouble to optimize `fillTables()` and its helper functions, as the code only runs a single time.

**Review by MichaelEischer — COMMENTED** [review]
I've so far only taken a look at the changes related to `NextSplitPoint()`. So far it looks like the only significant change is inlining `wpos = wpos % windowSize`, all other changes seem to have little to no effect on the generated assembly, see the comments for more details. 

I haven't taken a closer look at the polynomial calculations.

File: chunker.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ const (
 
 	// WindowSize is the size of the sliding window.
 	windowSize = 64
+	windowMask = windowSize - 1
```
[the remark this answers is not in the document]
Comment (wjiec) [reply chunker.go:14]: I tested using Go version: `go version go1.19.11 darwin/amd64` and found that there is a difference(same behavior on linux). I added the following code to the chunker.go file:
```go
func (c *BaseChunker) testAND() byte {
	return c.window[c.wpos&windowMask]
}

func (c *BaseChunker) testMod() byte {
	return c.window[c.wpos%(windowSize-1)]
}

func (c *BaseChunker) testMod1() byte {
	return c.window[c.wpos%63]
}
```
then I use the command `go tool compile -S chunker.go polynomials.go options.go > chunk.s` to view the compiled assembly instructions(abridged):
```asm
<unlinkable>.(*BaseChunker).testAND STEXT nosplit size=21 args=0x8 locals=0x0 funcid=0x0 align=0x0
	0x0000 00000 (chunker.go:236)	TEXT	<unlinkable>.(*BaseChunker).testAND(SB), NOSPLIT|ABIInternal, $0-8
	0x0000 00000 (chunker.go:237)	TESTB	AL, (AX)
	0x0002 00002 (chunker.go:237)	MOVQ	4208(AX), CX // wpos
	0x0009 00009 (chunker.go:237)	ANDL	$63, CX      // & 63
	0x000c 00012 (chunker.go:237)	MOVBLZX	4144(AX)(CX*1), AX
	0x0014 00020 (chunker.go:237)	RET

<unlinkable>.(*BaseChunker).testMod STEXT nosplit size=60 args=0x8 locals=0x0 funcid=0x0 align=0x0
	0x0000 00000 (chunker.go:240)	TEXT	<unlinkable>.(*BaseChunker).testMod(SB), NOSPLIT|ABIInternal, $0-8
	0x0000 00000 (chunker.go:241)	TESTB	AL, (AX)
	0x0002 00002 (chunker.go:241)	MOVQ	4208(AX), CX
	0x0009 00009 (chunker.go:240)	MOVQ	AX, DX
	0x000c 00012 (chunker.go:241)	MOVQ	$292805461487453201, AX
	0x0016 00022 (chunker.go:240)	MOVQ	DX, BX
	0x0019 00025 (chunker.go:241)	MULQ	CX
	0x001c 00028 (chunker.go:241)	ADDQ	CX, DX
	0x001f 00031 (chunker.go:241)	RCRQ	$1, DX
	0x0022 00034 (chunker.go:241)	SHRQ	$5, DX
	0x0026 00038 (chunker.go:241)	MOVQ	DX, SI
	0x0029 00041 (chunker.go:241)	SHLQ	$6, DX
	0x002d 00045 (chunker.go:241)	SUBQ	SI, DX
	0x0030 00048 (chunker.go:241)	SUBQ	DX, CX
	0x0033 00051 (chunker.go:241)	MOVBLZX	4144(BX)(CX*1), AX
	0x003b 00059 (chunker.go:241)	RET

<unlinkable>.(*BaseChunker).testMod1 STEXT nosplit size=60 args=0x8 locals=0x0 funcid=0x0 align=0x0
	0x0000 00000 (chunker.go:244)	TEXT	<unlinkable>.(*BaseChunker).testMod1(SB), NOSPLIT|ABIInternal, $0-8
	0x0000 00000 (chunker.go:245)	TESTB	AL, (AX)
	0x0002 00002 (chunker.go:245)	MOVQ	4208(AX), CX
	0x0009 00009 (chunker.go:244)	MOVQ	AX, DX
	0x000c 00012 (chunker.go:245)	MOVQ	$292805461487453201, AX
	0x0016 00022 (chunker.go:244)	MOVQ	DX, BX
	0x0019 00025 (chunker.go:245)	MULQ	CX
	0x001c 00028 (chunker.go:245)	ADDQ	CX, DX
	0x001f 00031 (chunker.go:245)	RCRQ	$1, DX
	0x0022 00034 (chunker.go:245)	SHRQ	$5, DX
	0x0026 00038 (chunker.go:245)	MOVQ	DX, SI
	0x0029 00041 (chunker.go:245)	SHLQ	$6, DX
	0x002d 00045 (chunker.go:245)	SUBQ	SI, DX
	0x0030 00048 (chunker.go:245)	SUBQ	DX, CX
	0x0033 00051 (chunker.go:245)	MOVBLZX	4144(BX)(CX*1), AX
	0x003b 00059 (chunker.go:245)	RET
```

File: chunker.go [hunk 3]
```diff
@@ -37,7 +38,7 @@ func init() {
 
 type chunkerState struct {
 	window [windowSize]byte
-	wpos   uint
+	wpos   uint64
```
[the remark this answers is not in the document]
Comment (wjiec) [reply chunker.go:41]: Yeah, I didn't consider that on a 32-bit system

File: chunker.go [hunk 4]
```diff
@@ -217,35 +213,24 @@ func (c *BaseChunker) NextSplitPoint(buf []byte) (int, uint64) {
...
-	c.wpos = (c.wpos + 1) % windowSize
-
-	digest = updateDigest(digest, c.polShift, &c.tables, b)
-	return digest
-}
-
-func appendByte(hash Pol, b byte, pol Pol) Pol {
-	hash <<= 8
-	hash |= Pol(b)
+	c.wpos++
```
[the remark this answers is not in the document]
Comment (wjiec) [reply chunker.go:234]: Indeed.

File: chunker.go (hunk 1 above)
Comment (wjiec) [reply chunker.go:135]: Indeed.

**Comment by wjiec** [comment]
inline `slide` into `reset` for greater performance gains:
```plain
goos: linux
goarch: amd64
pkg: github.com/restic/chunker
cpu: Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz
                    │    old.txt    │                 new.txt                 │
                    │    sec/op     │    sec/op      vs base                  │
ChunkerWithSHA256-8    189.3m ±  1%    184.3m ±  0%     -2.62% (p=0.000 n=10)
Chunker-8              79.32m ±  2%    74.19m ±  1%     -6.47% (p=0.000 n=10)
NewChunker-8           214.5µ ± 10%    125.3µ ± 46%    -41.55% (p=0.000 n=10)
PolDivMod-8            3.463n ±  1%   96.620n ±  0%  +2690.47% (p=0.000 n=10)
PolDiv-8               3.463n ±  1%   96.385n ±  0%  +2683.68% (p=0.000 n=10)
PolMod-8               3.463n ±  1%   96.380n ±  0%  +2682.73% (p=0.000 n=10)
PolDeg-8              0.3743n ±  0%   0.3748n ±  0%          ~ (p=0.382 n=10)
RandomPolynomial-8    2436.3µ ±  2%    363.6µ ±  2%    -85.08% (p=0.000 n=10)
PolIrreducible-8      1363.0µ ±  0%    127.6µ ±  0%    -90.64% (p=0.000 n=10)
PolMul-8                               121.5n ±  0%
PolGCD-8                               712.6n ±  0%
PolMulMod-8                            557.2n ±  0%
geomean                8.192µ          5.744µ          +75.87%

                    │   old.txt    │               new.txt               │
                    │     B/s      │     B/s       vs base               │
ChunkerWithSHA256-8   169.1Mi ± 1%   173.6Mi ± 0%  +2.69% (p=0.000 n=10)
Chunker-8             403.4Mi ± 2%   431.3Mi ± 1%  +6.92% (p=0.000 n=10)
geomean               261.2Mi        273.7Mi       +4.78%
```

File: chunker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply chunker.go:14]: `return c.window[c.wpos%63]` is wrong. it must be `return c.window[c.wpos%64]`. 

`return c.window[c.wpos&windowMask]` == `return c.window[c.wpos%(windowMask+1)]`.

[the remark this answers is not in the document]
Comment (MichaelEischer) [reply chunker.go:14]: Please use go 1.21 for the performance comparison. There's not much use in optimizing for no longer supported go versions.

**Comment by MichaelEischer** [comment]
Please split the polynomial-related changes into a separate PR. I'm currently pretty short of time, and I'd prefer to handle the changes in `chunker.go` first, as these are performance critical for restic in contrast to the polynomial changes.

**Comment by wjiec** [comment]
Okay, I'll be filing separate PRs later today.

**Pull request closed without merging**